components/brave_news/browser/brave_news_p3a.cc

```diff
@@ -67,110 +44,265 @@ void RecordGeneralUsage() {
+  if (total) {
+    subscription_counts_[subscribe_type] = *total;
+  }
+
+  const char* histogram_name;
+  switch (subscribe_type) {
+    case SubscribeType::kChannels:
+      histogram_name = kChannelCountHistogramName;
+      break;
+    case SubscribeType::kPublishers:
+      histogram_name = kPublisherCountHistogramName;
+      break;
+    default:
+      NOTREACHED();
+      return;
+  }
+
+  if (!IsMonthlyActiveUser()) {
+    // Only report for active users in the past month.
+    return;
+  }
+
+  if (subscription_counts_.contains(subscribe_type)) {
+    p3a_utils::RecordToHistogramBucket(histogram_name,
+                                       kSubscriptionCountBuckets,
+                                       subscription_counts_[subscribe_type]);
+  }
+}
+
+void NewsMetrics::RecordFeatureEnabledChange() {
+  bool enabled = IsNewsEnabled();
+  bool was_ever_enabled = prefs_->GetBoolean(prefs::kBraveNewsWasEverEnabled);
   if (!enabled && !was_ever_enabled) {
     // If the user clicked "no thanks" on the NTP, then we don't want to record
     // this as an opt-out, since they were never opted in.
     return;
   }
-  prefs->SetBoolean(prefs::kBraveNewsWasEverEnabled, true);
+  prefs_->SetBoolean(prefs::kBraveNewsWasEverEnabled, true);
   UMA_HISTOGRAM_BOOLEAN(kIsEnabledHistogramName, enabled);
 }
 
-void RecordAtInit(PrefService* prefs) {
-  RecordLastUsageTime(prefs);
-  RecordNewUserReturning(prefs);
+void NewsMetrics::RecordAtInit() {
+  RecordLastUsageTime();
+  RecordNewUserReturning();
+
+  RecordDirectFeedsTotal();
+  RecordWeeklyAddedDirectFeedsCount(0);
+  RecordWeeklySessionCount(false);
+  RecordWeeklyDisplayAdsViewedCount(false);
+
+  for (const auto action : kAllActionTypes) {
+    RecordTotalActionCount(action, 0);
+  }
+  RecordVisitCardDepth({});
+
+  if (IsNewsEnabled()) {
+    prefs_->SetBoolean(prefs::kBraveNewsWasEverEnabled, true);
+  }
+
+  report_timer_.Start(
+      FROM_HERE, base::Time::Now() + kReportInterval,
+      base::BindOnce(&NewsMetrics::RecordAtInit, base::Unretained(this)));
+}
+
+bool NewsMetrics::IsNewsEnabled() {
+  return prefs_->GetBoolean(prefs::kBraveNewsOptedIn) &&
+         prefs_->GetBoolean(prefs::kNewTabPageShowToday);
+}
 
-  RecordDirectFeedsTotal(prefs);
-  RecordWeeklyAddedDirectFeedsCount(prefs, 0);
-  RecordWeeklySessionCount(prefs, false);
-  RecordWeeklyDisplayAdsViewedCount(prefs, false);
-  RecordTotalCardViews(prefs, 0);
+bool NewsMetrics::IsMonthlyActiveUser() {
+  base::Time last_usage = prefs_->GetTime(prefs::kBraveNewsLastSessionTime);
+  return base::Time::Now() - last_usage < kMonthlyUserTimeThreshold;
+}
 
-  if (IsNewsEnabled(prefs)) {
-    prefs->SetBoolean(prefs::kBraveNewsWasEverEnabled, true);
+uint64_t NewsMetrics::AddToWeeklyStorageAndGetSum(const char* pref_name,
+                                                  int change) {
+  WeeklyStorage storage(prefs_, pref_name);
+  if (change > 0) {
```

Comment:
Just confirming we only ever want to increment or decrement by 1?

Reply:
yes